src/components/include/stacklist.h

```diff
@@ -0,0 +1,94 @@
+#ifndef STACKLIST_H
+#define STACKLIST_H
+
+/**
+ * Modified to support multi-core via a Treiber stack. This is not 100%
+ * a great solution as it isn't FIFO. However, we release *all*
+ * threads when unlocking, so the priority scheduling should take over
+ * at that point.
+ */
+
+#include <cos_component.h>
+#include <ps.h>
+
+struct stacklist {
+	thdid_t thdid;
+	struct stacklist *next;
+};
+
+struct stacklist_head {
+	struct stacklist *head;
+};
+
+static inline void
+stacklist_init(struct stacklist_head *h)
+{
+	h->head = NULL;
+}
+
+/*
+ * Remove a thread from the list that has been woken. Return 0 on
+ * success, and 1 if it could not be removed.
+ */
+static inline int
+stacklist_rem(struct stacklist *l)
+{
+	/*
+	 * Not currently supported with Trebor Stack. Threads that
+	 * wake early still have to wait their turn.
+	 */
+	return 1;
+}
+
+/* Add a thread that is going to block */
+static inline void
+stacklist_add(struct stacklist_head *h, struct stacklist *l)
```

Comment:
Allocating the struct stacklist *l in the function (on stack) may be better than just a pointer itself.